Create 2023-07-14-devx.md (#65)
input-output-hk/engineering: blog/2023-07-14-devx.md

---
slug: 2023-07-14-devx

> title: The IOG's Developer Experience Shell

Reviewer: This is strange punctuation. I think it should be

```suggestion
title: The IOG Developer`s Experience Shell
```

authors: [yvan]
tags: [devx, haskell.nix, hix, iogx, nix]
---

> # The IOG's Developer Experience Shell

Reviewer: same as above

```suggestion
# The IOG Developer's Experience Shell
```

The IOG's Developer Experience Shell, or `DevX` for short, is an opinionated development environment tailored for Haskell.
Built on top of Nix expressions, it provides a reproducible and seamless, full-featured developer shell with tools such as `cabal-install`, `ghc`, `hls`, and `hlint`.

The DevX shell's approach is a "one shell that fits all" solution for IOG's project needs.
This article aims to give you a technical deep dive on how to use it in your own Haskell project and how we managed such a substantial closure in a convenient way.

`DevX` is built upon `haskell.nix`, a framework for building Haskell packages with Nix.
In short, `haskell.nix` turns your Cabal or Stack project and its dependencies into a Nix expression.
We will explain how to use `haskell.nix` on its own using `hix`.

> We will conclude by introducing `iogx`, a Flake Template for Haskell Projects that we use at IOG.

Reviewer 2: After reading the intro I have questions:
- why do we need Nix? I.e. why can't we directly use cabal/stack?
- what is a flake?
- what is it that we're trying to fix?

As a non Nix user, I would probably not read the rest of the post because it seems like there isn't anything for me in it and I don't understand `haskell.nix`, `hix`, `Flake`.

So perhaps rephrase it with:
1. the problem we address with the DevX shell
2. the promise that we have a solution for it: read on to know more about it (hook the reader, but don't fail to deliver)
3. if you really want to mention Nix, perhaps mention that it's built upon the battle-tested `haskell.nix` but that knowing Nix isn't a prerequisite to read the post

> We will conclude by introducing `iogx`, a Flake Template for Haskell Projects that we use at IOG.

Reviewer: The structure should be:

### What
that is this line: `The IOG's Developer Experience Shell, or `DevX` for short, is an opinionated development environment tailored for Haskell.`

### Why
that is this line: `Built on top of Nix expressions, it provides a reproducible and seamless, full-featured developer shell with tools such as `cabal-install`, `ghc`, `hls`, and `hlint`.`

Then you should motivate its use. To do that you should state what the problem is that this solves. There are several levels of problem here: Org level: How do we sync workspaces to make sure we're all dealing with the same tech stack? Dev level: How do we make sure that Dev A can reuse the work of Dev B? Engineering level: How do guarantee that some behavior Dev A observes is not due to a difference in decencies, this removes an entire class of hypotheses we would need to investigate in this scenario.

 If it were me I would drive home that one of the things a program is, is a _semantic contract_. The program precisely and accurately states a meaning of some thing (the meaning of `2 + 2` _is_ the symbol `4`). Then taking this idea and running with it, in nix you get a semantic contract for a shell closure; for an environment (in programming languages we would call this a context). After that, you should `show, not tell`. In your motivation line you tell: `it provides a reproducible and seamless, full-featured developer shell with tools`, to show you should paint a picture:

> With this shell, all IOG developers have a  reproducible, full-featured and correct developer environment simply by invoking `nix develop` or whatever the invocation is. If its using direnv, then it should be: "simply by cd'ing to their working directory". No hassle, no fuss, and architecture independent. 

To show you want to evoke the image of the reader to seeing themselves hitting RET on `nix develop`.

### How
This is where you would talk more about nix and compare with the alternatives like cabal/stack etc. I would point out that stack has reproducible builds too, but is much more primitive than nix. You could also argue that it is much easier to add a non-haskell dependency in nix than in stack or cabal, and especially with flakes. 

### Where we are going
That's this line: `This article aims to give you a technical deep dive on how to use it in your own Haskell project and how we managed such a substantial closure in a convenient way.`. But you should also say "After reading this you should be able to ...."

The rest of the introduction is too much detail for an introduction:

> The DevX shell's approach is a "one shell that fits all" solution for IOG's project needs.

> `DevX` is built upon `haskell.nix`, a framework for building Haskell packages with Nix. In short, `haskell.nix` turns your Cabal or Stack project and its dependencies into a Nix expression. We will explain how to use `haskell.nix` on its own using `hix`.

Move this stuff to `DevX shell's philosophy` or a `What the DevX shell isn't` section.

## Getting started

> The `devx` GitHub repository provides a comprehensive [README](https://github.com/input-output-hk/devx) that documents the different outputs that `flake.nix` provides.

Reviewer 2: ```suggestion
The `devx` GitHub repository provides a comprehensive [README](https://github.com/input-output-hk/devx#readme) that documents the different outputs that `flake.nix` provides.
```

But, as this article assumes you might have not so much knowledge of how to deal with Nix, we will quickly discuss three ergonomic ways of using this tool with no Nix knowledge required.

> ### Direnv

Reviewer: The intro states:

> This article aims to give you a technical deep dive on how to use it in your own Haskell project and how we managed such a substantial closure in a convenient way.

So why are we immediately talking about `direnv`?  Its because of the last sentence in Getting Started. I think its better to talk about the IOG DevX shell first and then talk about ergonomic ways to use it. You could cut these for a part 2 blog post, that way you can limit the size of your audience: part 1 for nix people, part 2 for newcommers. If you want to keep it then it should be after you make good on the promise in the introduction, i,e., "how to use it in your own Haskell project". Once you do that, then you can say "BUT WAIT! Its _even_ better with direnv" (or lorri, or lorelei).

``` 
- name: Build
  uses: input-output-hk/actions/devx@latest
  with:
    platform: 'x86_64-linux'
    target-platform: '-windows'
    compiler-nix-name: 'ghc8107'
    minimal: false
    iog: true
- name: Build
  shell: devx {0}
  run: |
    cabal update
    cabal build
```

### VSCode DevContainer and GitHub CodeSpace

Finally, we also offer support for VSCode DevContainers and GitHub CodeSpaces.
To make the DevX developer shell available in a VSCode DevContainer or GitHub CodeSpace, simply add a file named `.devcontainer/devcontainer.json` with the following content:
```
{
   "image":"ghcr.io/input-output-hk/devx-devcontainer:ghc8107",
   "customizations":{
      "vscode":{
         "extensions":[
            "haskell.haskell"
         ],
         "settings":{
            "haskell.manageHLS":"PATH"
         }
      }
   }
}
```
You can follow the [Microsoft tutorial](https://code.visualstudio.com/docs/devcontainers/tutorial) to set-up your VSCode local DevContainer or you can give it a try by [opening a GitHub Codespace](https://codespaces.new/input-output-hk/cardano-base?quickstart=1) on [`cardano-base`](https://github.com/input-output-hk/cardano-base) repository!

> ## Benchmarking

Reviewer: This should also be after you explain the shell. Also since you're comparing to `nix develop` you are assuming some knowledge of nix from your audience.

To showcase the efficiency and speed that our download closure hack brings compared to a call to `nix develop`, we'll present a short benchmarking comparison.

What isn't measured is the time for evaluating the closure, compressing it, and uploading it to `ghcr.io`, as it's done by our CI and represents the caching part of the process.
We measure the speed of downloading, decompressing, and performing a `nix import` on the closure against the speed of `nix develop`, so we essentially save the time of the evaluation of the Nix expression:

**TODO: add the benchmarks ...**

## Hix

Hix is a command-line tool designed to facilitate the addition of `haskell.nix` support to existing Haskell projects.

> You need Nix installed and configured with `experimental-features = [ "nix-command" "flakes" ];` (details can be found at the [NixOS Wiki](https://nixos.wiki/wiki/Flakes)).

Reviewer: Good!

### Initializing Hix and Using Nix

> The `hix init` command is used to add a `flake.nix` and `nix/hix.nix` file to the project.

Reviewer: use an active voice:

```suggestion
The `hix init` command adds a `flake.nix` and `nix/hix.nix` file to the project root.
```

Once this is done, you can utilize regular Nix tools.

> For instance, to run `cabal build` on the `hello` package from Hackage, you would execute the following:

Reviewer: Good! It is good pedagogy to show how you would do it a familiar way _before_ explaining the new way.

```shell
cabal unpack hello
cd hello-1.0.0.2
nix run "github:input-output-hk/haskell.nix#hix" -- init
nix develop
cabal build

> ```

Reviewer: explain this example. If you show any code, then you should explain it before making your point with it. In this case, that would be something like this:

> As you can see, we use cabal to pull hello from hackage (`cabal unpack hello`), then we enter the directory and run `hix init`. Next we utilize the `flake` we added via `hix init` to enter a development shell with `nix develop`. Now that we are in the development shell we can use cabal as one normally would.

Two other points:
- You could use `cabal list --installed` to show the difference between environments before `nix develop` and after.
- You use `nix run "<hix-url>" -- init`, which is fine _for people who know nix_. So you are again assuming nix familiarity with you audience. If you want to keep this line then make sure you introduce it in the hix section so a reader isn't confused when they get here.

> You can view the contents of the flake using `nix flake show`.

Reviewer: Show this! Don't make your audience do work.

> To build a component with Nix, use `nix build .#hello:exe:hello`, and to build and run a component, `nix run .#hello:exe:hello` would be the command of choice.

Reviewer: aha! this should be earlier and in the hix section. Perhaps you add a `Ultra fast nix primer` section after the Getting Started section.

### Installing Hix

Hix can be installed with the following command:

```shell
nix-env -iA hix -f https://github.com/input-output-hk/haskell.nix/tarball/master
```

> To update Hix to the latest version, simply run `hix update`.

Reviewer: another good place to explain the `nix run <hix-url>`.

### Using Hix Commands

> Commands like `hix develop`, `hix flake`, `hix build`, and `hix run` work similarly to their Nix counterparts.

Reviewer: Assumes nix knowledge

However, instead of utilizing the `flake.nix`, a boilerplate Haskell.nix `flake.nix` file is added to `.hix-flake/flake.nix`.

> This approach can be beneficial if the project already includes a `flake.nix` or if there's no intention to maintain one.

Reviewer: why is it beneficial if there is already a `flake.nix`? If one already exists then what is my motivation for using `hix`? This is a genuine question, not a comment on the writing :)

Once this step is done, you can execute commands without the need for `hix init`:

```shell
hix develop
hix flake show
hix build .#hello:exe:hello
hix run .#hello:exe:hello
```

### Using `hix-shell` and `hix-build`

The `hix-shell` and `hix-build` commands emulate the behaviour of `nix-build` and `hix-shell` if a boilerplate `default.nix` and `shell.nix` were present.
These commands are executed as follows:
```shell
hix-shell --run 'cabal build all'
hix-build -A hsPkgs.hello.components.exes.hello
```

> Hix streamlines the use of Haskell.nix, offering an easier way to integrate it into your existing Haskell projects.

Reviewer: Aha! This is the important bit. This is the motivation for all the hix stuff that came before it. You should move this before all the code examples. As a reader who uses nix everyday I want to know what `hix` buys me over regular flaked nix. Ideally you would have an example that says "See, with hix this happens or that happens or it does this for you".

> ## IOGX: A Flake Template for Haskell Projects at IOG

Reviewer: I think the IOGX stuff should be part 3. 

And for motivation you should say "the learning curve for nix is steep and the documentation is not beginner friendly" and then cite a bunch of reddit or discourse threads or the nix documentation working group problem statement.
